webapp/models.py

```diff
@@ -112,20 +136,44 @@ def process_html(self, html):
 
         return str(soup)
 
-    def get_document(self, path):
+    def _parse_document_topic(self, topic):
         """
-        Retrieve and return relevant data about a document:
-        - Title
-        - HTML content
-        - Navigation content
+        Parse a topic object retrieve from Discourse
+        and return document data:
+        - title: The title
+        - body_html: The HTML content of the initial topic post
+                     (with some post-processing)
+        - updated: A human-readable data, relative to now
+                   (e.g. "3 days ago")
+        - forum_link: The link to the original forum post
         """
 
-        document, nav_html = self.get_frontpage()
+        updated_datetime = dateutil.parser.parse(
+            topic["post_stream"]["posts"][0]["updated_at"]
+        )
+
+        return {
+            "title": topic["title"],
+            "body_html": self._process_html(
+                topic["post_stream"]["posts"][0]["cooked"]
+            ),
+            "updated": humanize.naturaltime(
+                updated_datetime.replace(tzinfo=None)
+            ),
+            "forum_link": f"{self.base_url}/t/{topic['slug']}/{topic['id']}",
+        }
+
+    def _get_topic(self, path):
+        """
+        Retrieve topic object by path
+        """
 
-        if f"/t/{path}" != document["path"]:
-            topic = self.get_topic(path)
-            document = self.parse_topic(topic)
+        response = self.session.get(
+            f"{self.base_url}/t/{path}.json", allow_redirects=False
+        )
+        response.raise_for_status()
 
-        document["body_html"] = self.process_html(document["body_html"])
+        if response.status_code >= 300:
+            raise RedirectFoundError(response=response)
 
-        return document, nav_html
+        return response.json()
```

Comment:
some errors are not handled in this function (5xx, 4xx, json decode errors)

Reply:
For the time being, I'm completely comfortable with that.

If my class encounters a `500` or `4xx` response from requests, it will bubble up a [requests exception](http://docs.python-requests.org/en/master/_modules/requests/exceptions/). These errors perfectly explain what went wrong - probably better than if I were to mask it by raising my own error. I don't think raising my own error would add any value.

If it encounters invalid JSON then it will bubble up a `JSONDecodeError`. This is slightly more obscure - it might be nice to explain to the client that it was the the Discourse forum that sent invalid JSON. But the `JSONDecodeError` is still pretty informative and certainly gives you more than enough of a thread to pull to fix your problem.

There are plenty of other errors I'm also not catching, to do with the exact structure of the JSON dictionaries I'm expecting etc. And I also don't think it's particularly high priority to write custom errors for that either - again the natural errors they produce will be informative and give you a fairly clear debugging path.

So for the time being, I'm going to consider catching such exceptions out of scope for this PR and probably for the foreseeable future. As and when we discover particularly painful obscure errors, we can, should, and I believe, will catch them and expose them in a clearer fashion.

Reply:
I just mention this since we already encoutered it on snapcraft.io , when we can we should add sentry to catch errors.